aws-ts-ruby-on-rails/init.ts

```diff
@@ -0,0 +1,112 @@
+import * as fs from "async-file";
+import * as mustache from "mustache";
+
+type Init = {[config: string]: InitConfig};
+
+interface InitConfig {
+    files?: {[path: string]: InitFile};
+    packages?: {[manager: string]: InitPackages};
+    commands?: {[command: string]: InitCommand};
+    services?: {[manager: string]: InitServices};
+}
+
+interface InitFile {
+    content: Promise<string>;
+    mode: string;
+    owner: string;
+    group: string;
+}
+
+// TODO: support full syntax for packages.
+type InitPackages = string[];
+
+interface InitCommand {
+    command: string;
+    test?: string;
+}
+
+type InitServices = {[service: string]: InitService};
+
+interface InitService {
+    enabled?: boolean;
+    ensureRunning?: boolean;
+}
+
+// createUserData produces a single user data script out of the given init structure.
+export async function createUserData(configs: string[], init: Init): Promise<string> {
+    let script = `#!/bin/bash -xe\n\n`;
```

Comment:
This appears to be taking a cloud-init like schematic as input and lowering it to bash. You should be able to just generate cloud-init format directly and pass that as userData.

Reply:
I spent a little time looking into this. Unless I'm missing something, cloud-init isn't an identical 1-for-1 match. It does support all the same capabilities, however (minus the package management commands, again unless I'm missing some extensibility point (which is likely)).

My goal was to make it easy to convert existing CloudFormation. For that reason, I wonder if it makes sense to support the types that I have in here, and simply lower to cloud-init. There are several reasons why doing cloud-init directly actually kind of stinks: the MIME multi-part encoding required for multi-file configs, and the odd merging behavior it has for duplicate dictionaries (this example is a good one where you want to have multiple conceptual steps).

I agree, lowering to Bash is as gross as it gets, so cloud-init seems better.